src/main/java/com/o19s/es/ltr/ranker/parser/LightGBMJsonParser.java

```diff
@@ -0,0 +1,275 @@
+            PARSER = new ObjectParser<>("split", SplitParserState::new);
+
+            PARSER.declareBoolean(SplitParserState::setDefaultLeft, new ParseField("default_left"));
+            PARSER.declareString(SplitParserState::setMissingType, new ParseField("missing_type"));
+            PARSER.declareFloat((p, s) -> {}, new ParseField("split_gain"));
+            PARSER.declareInt((p, s) -> {}, new ParseField("internal_value"));
+            PARSER.declareInt((p, s) -> {}, new ParseField("internal_count"));
+            PARSER.declareString(SplitParserState::setDecisionType, new ParseField("decision_type"));
+            PARSER.declareInt((p, s) -> {}, new ParseField("split_index"));
+            PARSER.declareFloat(SplitParserState::setThreshold, new ParseField("threshold"));
+            PARSER.declareInt(SplitParserState::setSplitFeature, new ParseField("split_feature"));
+            PARSER.declareObject(SplitParserState::setLeftChild, SplitParserState::parse, new ParseField("left_child"));
+            PARSER.declareObject(SplitParserState::setRightChild, SplitParserState::parse, new ParseField("right_child"));
+
+            PARSER.declareFloat(SplitParserState::setLeafValue, new ParseField("leaf_value"));
+            PARSER.declareInt((p, s) -> {}, new ParseField("leaf_count"));
+            PARSER.declareInt((p, s) -> {}, new ParseField("leaf_index"));
+        }
+
+        // Properties found in a split
+        private Boolean defaultLeft;
+        private MissingType missingType;
+        private String decisionType;
+        private float threshold = Float.NaN;
+        private int splitFeature = -1;
+        private SplitParserState rightChild;
+        private SplitParserState leftChild;
+
+        // Properties found in a leaf
+        private Float leafValue;
+
+        public static SplitParserState parse(XContentParser parser, FeatureSet set) {
+            SplitParserState split = PARSER.apply(parser, set);
+            if (split.isSplit()) {
+                if (!split.hasAllSplitFields()) {
+                    throw new ParsingException(parser.getTokenLocation(), "This split does not have all the required fields");
+                }
+                if (!"<=".equals(split.decisionType)) {
+                    throw new ParsingException(parser.getTokenLocation(), "Only the <= decision_type is supported");
+                }
+                if (!split.defaultLeft) {
+                    throw new ParsingException(parser.getTokenLocation(), "default_left must be true");
+                }
+            }
+            return split;
+        }
+
+        // This is the isZero check used in lightgbm.
+        static final float ZERO_THRESHOLD = 1e-35F;
+
+        static boolean isZero(float fval) {
+            return fval > -ZERO_THRESHOLD && fval <= ZERO_THRESHOLD;
+        }
+
+        // See lightgbm src/io/tree.cpp Tree::NumericalDecisionIfElse
+        // TODO: NaN's are invalid lucene scores, should we even check?
```

Comment:
well actually it depends on what's done on the FeatureVector#reset() method, basically missing values are not set so if the FeatureVector decides to initialize everything to NaN then that it's the missing value indicator...
On the other hand we have no FeatureVector implementations that reset to NaN everything uses 0 int reset() currently.
It's really not obvious to decide where this logic should belong to...

Reply:
I think since our underlying stack is lucene and lucene defines scores as positive and non-NaN we should expect the same. reset() can be documented to reset to 0 and that NaN is invalid. This seems like the most obvious mental model and can be easily described.